exec/java-exec/src/test/java/org/apache/drill/exec/udf/dynamic/TestDynamicUDFSupport.java

```diff
@@ -104,8 +104,10 @@ public static void buildAndStoreDefaultJars() throws IOException {
   @Before
   public void setupNewDrillbit() throws Exception {
     udfDir = dirTestWatcher.makeSubDir(Paths.get("udf"));
+    File udfLocalDir = dirTestWatcher.makeSubDir(Paths.get("udf", "local"));
     Properties overrideProps = new Properties();
     overrideProps.setProperty(ExecConstants.UDF_DIRECTORY_ROOT, udfDir.getAbsolutePath());
+    overrideProps.setProperty(ExecConstants.UDF_DIRECTORY_LOCAL, udfLocalDir.getAbsolutePath());
```

Comment:
We've got lots of local directory properties. Hard to keep them all in sync. I wonder if we can use a feature of HOCON to default them to a known structure:

```
exec: {
   ...
   local: {
     baseDir: "/tmp/drill",
     udfDir: "${drill.exec.local.baseDir}/udf",
     pluginDir: "${drill.exec.local.baseDir}/plugins",
     ...
},
```

Probably some setup to do in the `ClusterFixture` and `DirTestWatcher` to get everything set up.

Reply:
Could you please clarify it a little bit?
Are you proposing to create one more test conf file with these configs or updating one of the existing ones?
In the first case, I'm afraid we can face up with the problem connected with the order of reading conf files: the same config values would depend on the order of reading config files.
In the second case, it may affect other tests.

Reply:
@vvysotskyi, glad you were able to simplify the test. I realize you didn't write them, so I appreciate taking the time to clean them up.

The random thought on the directories was that we have 4 or 5 (more) local directories that we set up. It seems we configure each separately. For example, from `drill-module.conf`:

```
drill.tmp-dir: "/tmp"
drill.tmp-dir: ${?DRILL_TMP_DIR}
...
  sys.store.provider: {
    local: {
      path: "/tmp/drill",
    }
  trace: {
    directory: "/tmp/drill-trace",
    filesystem: "file:///"
  },
  tmp: {
    directories: ["/tmp/drill"],
    filesystem: "drill-local:///"
  },
  compile: {
    code_dir: "/tmp/drill/codegen"
...
  spill: {
    // *** Options common to all the operators that may spill
    // File system to use. Local file system by default.
    fs: "file:///",
    // List of directories to use. Directories are created
    // if they do not exist.
    directories: [ "/tmp/drill/spill" ]
...
  udf: {
    directory: {
      // Base directory for remote and local udf directories, unique among clusters.
```

And probably more. To move where Drill stores temp files, the user must change all of these properties.

Fortunately, @arina-ielchiieva did a nice job with the UDF directories: they all are computed from the base directory:

```
   directory: {
      // Base directory for remote and local udf directories, unique among clusters.
      base: ${drill.exec.zk.root}"/udf",

      // Path to local udf directory, always created on local file system.
      // Root for these directory is generated at runtime unless Drill temporary directory is set.
      local: ${drill.exec.udf.directory.base}"/udf/local",

      // Set this property if custom file system should be used to create remote directories, ex: fs: "file:///".
      // fs: "",
      // Set this property if custom absolute root should be used for remote directories, ex: root: "/app/drill".
      // root: "",

      // Relative path to all remote udf directories.
      // Directories are created under default file system taken from Hadoop configuration
      // unless ${drill.exec.udf.directory.fs} is set.
      // User home directory is used as root unless ${drill.exec.udf.directory.root} is set.
      staging: ${drill.exec.udf.directory.base}"/staging",
      registry: ${drill.exec.udf.directory.base}"/registry",
      tmp: ${drill.exec.udf.directory.base}"/tmp"
    }
```

So, my question was, can we do the same thing for all the other local directories? Allow each to be custom-set, but default them to be computed from a single base directory. That way, if a unit test or install wants to move the Drill local directories to, say, `/var/drill/tmp`, they only need change a single config line and everything else follows automatically.

This can be done in the existing conf file as was done for UDFs. And, I guess to preserve compatibility, we'd have to leave the properties where they are; we'd just change their values.

If you feel this is a bit out of scope for this PR then we can file a Jira for the change.

```

Reply:
@paul-rogers, thanks for the explanation and pointing to some places of `drill-module.conf`.
I have also looked at it and found one more place:
```
  zk: {
    connect: "localhost:2181",
    root: "drill",
```
`drill.exec.zk.root` is used in several places as a base directory, and I think most of the users use its default value, so changing it may cause problems. Regarding other places, it would be easier to update their values without affecting users, but I think it would be better to do it in the scope of another Jira.

Reply:
The `zk.root` path is a path on ZK, not the file system. So, my thought is that the ZK path would not be set as part of the proposed system.

I do agree that we should tackle any fix as another PR. Filed DRILL-7593 to track the issue.

Reply:
But in some places, it is used as a path on the file system, so I mentioned it. Thanks for logging the new Jira ticket!

Reply:
Thanks for the explanation. Since that field is supposed to be relative to the ZK root, it probably should not also be used as a file system path. I suspect that, somewhere, sometime, somebody got confused.

If you know where those usages are, please add a note to the JIRA ticket; we should fix those at the same time to us some other config variable for the file system.